### PR TITLE
SSD-Mobilenet-v2 TensorRT Generation from Frozen Graph using UFF Parser

### LINKED ISSUE #89: Object detector retraining and deployment for Jetson using TLT & TensorRT

### PR DESCRIPTION
**Pull request by undefined-references**
I have added a docker container to install required packages and run a script to generate a TensorRT Engine as I explained in #89 on Jetson Nano

### CODE REVIEW

**Comment by mhejrati** [comment]
@JsonSadler do you mind reviewing this?

**Comment by JsonSadler** [comment]
> @JsonSadler do you mind reviewing this?

Hi @mhejrati 
Sure.

File: training/tensorrt_generation/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# docker can be installed on the dev board following these instructions:
+# https://docs.docker.com/install/linux/docker-ce/debian/#install-using-the-repository , step 4: arm64
+# 1) build: docker build -f Dockerfile -t "neuralet/l4t-tensorrt-conversion" .
```
Comment (mhejrati) [inline training/tensorrt_generation/Dockerfile]: How about this?
neuralet/jetson-nano:tf-ssd-to-trt

File: training/tensorrt_generation/build_engine.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        "MultipleGridAnchorGenerator/Concatenate": concat_priorbox,  # for 'ssd_mobilenet_v1_coco'
+        "Concatenate": concat_priorbox,  # for other models
+        "concat": concat_box_loc,
+        "concat_1": concat_box_conf
+    }
+
+    graph.collapse_namespaces(namespace_plugin_map)
+    graph.remove(graph.graph_outputs, remove_exclusive_dependencies=False)
+    graph.find_nodes_by_op("NMS_TRT")[0].input.remove("Input")
+    if model == 'ssd_mobilenet_v1_coco':
```
Comment (mhejrati) [inline training/tensorrt_generation/build_engine.py]: What does this do?

File: training/tensorrt_generation/build_engine.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    """Replace all 'FusedBatchNormV3' in the graph with 'FusedBatchNorm'.
+    NOTE: 'FusedBatchNormV3' is not supported by UFF parser.
+    https://devtalk.nvidia.com/default/topic/1066445/tensorrt/tensorrt-6-0-1-tensorflow-1-14-no-conversion-function-registered-for-layer-fusedbatchnormv3-yet/post/5403567/#5403567
+    """
+    for node in graph.find_nodes_by_op('FusedBatchNormV3'):
+        gs.update_node(node, op='FusedBatchNorm')
+    return graph
+
+def parse_gridAnchor(graph):
+
```
Comment (mhejrati) [inline training/tensorrt_generation/build_engine.py]: docstring, please :)

File: training/tensorrt_generation/config.ini [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[MODEL]
```
Comment (mhejrati) [inline training/tensorrt_generation/config.ini]: Can we add documentation for each of these configs?

**Review by mhejrati — CHANGES_REQUESTED** [review]

File: training/tensorrt_generation/Dockerfile (hunk 1 above)
Comment (undefined-references) [reply training/tensorrt_generation/Dockerfile]: modified, thanks

File: training/tensorrt_generation/build_engine.py (hunk 3 above)
Comment (undefined-references) [reply training/tensorrt_generation/build_engine.py]: added, thanks

File: training/tensorrt_generation/build_engine.py (hunk 2 above)
Comment (undefined-references) [reply training/tensorrt_generation/build_engine.py]: This line removes a connection from image_tensor node to the Input node and this connection just exists on ssd-mobilenet-v1. Modified this `if` to not to check it by the model name as it's not relevant by just renaming it.

File: training/tensorrt_generation/config.ini (hunk 4 above)
Comment (undefined-references) [reply training/tensorrt_generation/config.ini]: yes, I also added some sample configs.

**Comment by undefined-references** [comment]
Hi, I modified the pr contents as bellow:
- Modularized the script to support both ssd-mobilenet-v1 and ssd-mobilenet-v2 models which are trained using any version of object detection api and in a way to be able to add support for ssd-inception soon.
- Added different sample config files for both v1 and v2 with documentations 
- Moved files into ssd subdirectory
- Removed download_model script and added download path to config files and moved downloading part to the main script 
- Performed required Dockerfile and Readme modifications
@mhejrati @JsonSadler

**Review by mhejrati — APPROVED** [review]
Thanks for making the changes.

**Pull request merged**